client/cli/src/logger.rs

```diff
@@ -0,0 +1,248 @@
+pub fn init_logger(pattern: &str, log_rotation_opt: &LogRotationOpt) -> Result<()> {
+
+	let mut builder = LogSpecBuilder::new();
+	// Disable info logging by default for some modules:
+	builder.module("ws", log::LevelFilter::Off);
+	builder.module("yamux", log::LevelFilter::Off);
+	builder.module("hyper", log::LevelFilter::Warn);
+	builder.module("cranelift_wasm", log::LevelFilter::Warn);
+	// Always log the special target `sc_tracing`, overrides global level
+	builder.module("sc_tracing", log::LevelFilter::Info);
+	// Enable info for others.
+	builder.default(log::LevelFilter::Info);
+
+	// Add filtesr defined by RUST_LOG.
+	builder.insert_modules_from(LogSpecification::env()?);
+
+	// Add filters passed in as argument.
+	builder.insert_modules_from(LogSpecification::parse(pattern)?);
+
+	// Build the LogSpec.
+	let spec = builder.build();
+
+	// Use timestamps to differentiate logs.
+	let naming = Naming::Timestamps;
+	// Never cleanup old logs; let the end-user take care of that.
+	let cleanup = Cleanup::Never;
+
+	let age = log_rotation_opt.log_age;
+	let size = log_rotation_opt.log_size;
+
+	// Build a Criterion from the options.
+	let criterion = match (age, size) {
+		(Some(a), None) => Criterion::Age(a),
+		(None, Some(s)) => Criterion::Size(s),
+		// Default to rotating with a size of `DEFAULT_ROTATION_SIZE`.
+		(None, None) => Criterion::Size(DEFAULT_ROTATION_SIZE),
+		_ => return Err(Error::Input("Only one of Age or Size should be defined".into()))
+	};
+
+	let isatty = atty::is(atty::Stream::Stderr);
+	let logger = Logger::with(spec)
+		.format(file_fmt)
+		.format_for_stderr(colored_fmt)
+		.format_for_stdout(colored_fmt)
+		.rotate(criterion, naming, cleanup);
+
+	let logger = match (log_rotation_opt.log_directory.as_ref(), isatty) {
+		// Only log to stderr using colored format; nothing to file, nothing to stdout.
+		(None, true) => logger.log_target(LogTarget::StdErr),
+		// Log to stderr using file format, log to stdout using colored format.
+		(None, false) => logger
+			.log_target(LogTarget::DevNull)
+			.format_for_stderr(file_fmt)
+			.duplicate_to_stderr(Duplicate::All)
+			.duplicate_to_stdout(Duplicate::Info),
+		// Log to stderr with colored format, log to file with file format. Nothing to stdout.
+		(Some(file), true) => logger.log_target(LogTarget::File)
+			.duplicate_to_stderr(Duplicate::All)
+			.directory(file),
+		// Log to stderr with file format, log to file with file format, log to stdout with colored format.
+		(Some(file), false) => logger
+			.log_target(LogTarget::File)
+			.format_for_stderr(file_fmt)
+			.duplicate_to_stderr(Duplicate::All)
+			.duplicate_to_stdout(Duplicate::Info)
+			.directory(file),
+	};
+
+	logger.start().map_or_else(|_| ().map_err(|e| e.into())
+}
+
+fn kill_color(s: &str) -> String {
+	lazy_static! {
+		static ref RE: Regex = Regex::new("\x1b\\[[^m]+m").expect("Error initializing color regex");
+	}
+	RE.replace_all(s, "").to_string()
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
This is minimal testing. Perhaps we could run substrate for like 30 seconds and then do some assertions on whether or not data has been written to stdout/stderr/log_directory ?

Reply:
That would be a test for flexi_logger but not for us. As "user" we expect that the library does its job